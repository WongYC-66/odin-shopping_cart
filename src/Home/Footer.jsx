import styles from "./footer.module.css"

function Footer() {

  return (
    <footer>
      <p>DISCLAIMER - This is a fake shopping website built with React.</p>
      Designed and coded by YC Wong @2023
    </footer>
  )
}

export default Footer
