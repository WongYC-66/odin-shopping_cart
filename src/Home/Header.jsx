import { Link } from "react-router-dom";
import styles from "./header.module.css"
import StoreLogo from './store-solid.svg';
import UserLogo from './user-solid.svg';
// import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
// import { icon } from '@fortawesome/fontawesome-svg-core/import.macro'

function Header({ cart = [] , currentPage="home", handleToggleCart}) {
  // console.log(items)
  // console.log(currentPage)
  const loginClick = () => {
    console.log(currentPage)
    if(currentPage === 'home')
      alert('Please go to store page first.')
    else
      alert('Login feature under construction.')
  }

  let total = 0;
  if (cart.length > 0) {
    total = cart.map(x => x.count).reduce((x, sum) => x + sum, 0);
  }

  return (
    <header >
      <h1>YUMEKO</h1>
      <div className={styles.navBar}>
        <li><Link to="/">Home</Link></li>
        <li><Link to="/store">Store</Link></li>
      </div>
      <div className={styles.icon}>
        <div className={styles.cartIcon}>
          <img src={StoreLogo} alt="StoreLogo" onClick={handleToggleCart}/>
          {cart.length > 0 ? <div title="cartCount" className={styles.count}>{total}</div> : ''}
        </div>
        <img src={UserLogo} alt="UserLogo" onClick = {loginClick} />
      </div>
    </header>
  )
}

export default Header
