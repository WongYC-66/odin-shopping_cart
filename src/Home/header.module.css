header, .icon, .icon img{
    display: flex;
    align-items: center;
    flex-direction: row;
    justify-content: space-between;
}

header li {
    list-style-type: none;
    padding: 15px 30px;
}

header a{
    text-decoration: none;
    color: black;
}

.navBar{
    display: flex;
    gap: 35%;
    font-size: 1.2rem;
}

h1{
    font-size: 3rem;
}

.icon img{
    width: 35px;
}

.icon{
    gap: 30px;
}

.icon .cartIcon{
    display: flex;
}

.cartIcon .count{
    color: darkorange;
    font-weight: 600;
    margin-left: 5px;
}

/* hover */
header li:hover, header li:hover > a{
    background-color: black;
    color: whitesmoke;
    cursor: pointer;
}

.icon img:hover{
    transform: scale(1.1);
    transition: transform 0.2s ease-in-out;
    cursor: pointer;
}
