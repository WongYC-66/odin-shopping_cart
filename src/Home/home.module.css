*{
    padding: 0;
    margin: 0;
    box-sizing: border-box;
}

body{
    font-family: Arial, Helvetica, sans-serif;
    text-align: center;
}

.container {
    padding: 15px;
    background-color: whitesmoke;
    display: grid;
    align-content: space-between;
    /* grid: 10px 500px 30px; */
    height: 100vh;
}


main *{
    padding:  15px;
}

main button {
    padding: 15px 30px;
    background-color: black;
    color: whitesmoke;
    font-weight: 800;
    border: none;
}

main button:hover{
    background-color: yellow;
    color: black;
    cursor: pointer;
}
