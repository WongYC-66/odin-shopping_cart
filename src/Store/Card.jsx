// import { useState, useEffect } from 'react'
import styles from './card.module.css'

function Card({ item, onChange, addToCart }) {
    // console.log(onChange)

    return (
        <div className={styles.card}>
            <img src={item.image} alt={item.title} />
            <h5>{item.title}</h5>
            <h4>{item.price} $</h4>
            <div className={styles.cardControl}>
                <button className = {styles.minusBtn} onClick={() => onChange(item.id, -1)}>-</button>
                <div className = {styles.countText}>{item.count}</div>
                <button className = {styles.plusBtn} onClick={() => onChange(item.id, +1)}>+</button>
                <button className = {styles.addToCartBtn} onClick={() => addToCart(item.id)}>Add to Cart</button>
            </div>
        </div>
    )
}

export default Card
