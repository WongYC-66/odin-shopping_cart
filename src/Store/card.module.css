.card img{
    height: 100px;
    /* width: 100px; */
    aspect-ratio: 1;
}

.card {
    border : 4px solid lightgrey;
    border-radius: 25px;
    background-color: white;
    padding: 10px;
    /*  */
    /* display: flex; */
    /* flex-direction: column; */
    display: grid;
    grid: 2fr 1fr 1fr 1fr / auto;
    gap: 3px;
    justify-items: center;
    align-items: center;
}

.card > .cardControl{
    display: flex;
    gap: 10px;
    align-items: center;
}

.cardControl .minusBtn, .cardControl .plusBtn{
    background-color: orange;
    border: 1px solid ghostwhite;
    padding: 1px 15px;
    border-radius: 20px;
    color: whitesmoke;
    font-weight: 800;
}

.cardControl .countText{
    /* background-color: aqua; */
    color: darkorange;
    font-weight: 800;
}

.cardControl .addToCartBtn{
    background-color: darkorange;
    border: none;
    color: whitesmoke;
    font-weight: 600;
    padding: 5px 10px;
}
.card > h5 {
    overflow: hidden;
}

.cardControl button:hover{
    background-color: green;
    color: whitesmoke;
    transform: translate(3px);

}

.cardControl button:active{
    transform: translate(5px);
    outline: none;
}
