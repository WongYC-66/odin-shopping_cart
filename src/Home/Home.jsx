import { useNavigate } from "react-router-dom"
import { useState, useEffect } from 'react'
import styles from "./home.module.css"
import Header from './Header.jsx'
import Footer from './Footer.jsx'
import ClosePhoto from './closet.webp';
import Cart from '../Cart/Cart.jsx'


function Home() {

  const [cart, setCart] = useState([]);
  const [showCart, setShowCart] = useState(false);

  const navigate = useNavigate();
  const navigateToStore = () => {
    // 👇️ navigate to /store
    navigate('/store');
  };

  const handleToggleCart = () => {
    if (showCart)
      setShowCart(false)
    else
      setShowCart(true)
  }

  const handleCartCountChange = (id, number) => {
    let newCartItems = [...cart]
    newCartItems.forEach(item => {
      if (item.id === id) item.count += number;
    })
    newCartItems = newCartItems.filter(item => item.count > 0)
    setCart(newCartItems)
    localStorage.setItem('cart', JSON.stringify(newCartItems));
  }

  useEffect(() => {
    // get history cart 
    const oldCart = JSON.parse(localStorage.getItem('cart'));
    if (oldCart) {
      setCart(oldCart)
    }
  }, []);

  return (
    <div className={styles.container}>
      <Header cart={cart} currentPage={"home"} handleToggleCart={handleToggleCart} />

      <main>
        <img src={ClosePhoto} alt="closetPhoto" />
        <h1 id="title">THE LONG WEEKEND SALE</h1>
        <h2>{`Fashion Clothes - Q'3`}</h2>
        <p>Grab it while stock last</p>
        <button onClick={navigateToStore}>SHOP NOW</button>
      </main>

      {showCart && <Cart cart={cart}
        handleToggleCart={handleToggleCart}
        handleCartCountChange={handleCartCountChange}
      />
      }

      <Footer />
    </div>
  )
}

export default Home
