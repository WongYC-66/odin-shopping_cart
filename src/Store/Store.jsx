import { useState, useEffect } from 'react'
import styles from "./store.module.css"
import Header from '../Home/Header.jsx'
import Footer from '../Home/Footer.jsx'
import Card from './Card.jsx'
import Cart from '../Cart/Cart.jsx'


function Store({testItems} = []) {

    const [items, setItems] = useState([]);
    const [cart, setCart] = useState([]);
    const [showCart, setShowCart] = useState(false);

    // console.log(cart)
    // console.log(testItems)


    const handleToggleCart = () => {
        if (showCart)
            setShowCart(false)
        else
            setShowCart(true)
    }

    const handleCountChange = (id, number) => {
        let newItems = [...items]
        newItems.forEach(item => {
            if (item.id === id) item.count += number;
            if (item.count < 0) item.count = 0;
        })
        setItems(newItems)
    }

    const handleCartCountChange = (id, number) => {
        let newCartItems = [...cart]
        newCartItems.forEach(item => {
            if (item.id === id) item.count += number;
        })
        newCartItems = newCartItems.filter(item => item.count > 0)
        setCart(newCartItems)
        localStorage.setItem('cart', JSON.stringify(newCartItems));
    }

    const handleAddToCart = (id) => {
        let storeItem = items.find(x => x.id === id)
        if (storeItem.count === 0) return;
        let newCart = [...cart]
        let cartItem = newCart.find(x => x.id === id)
        if (cartItem) {
            //already in the cart, just + the count, then update cart
            cartItem.count += storeItem.count
        } else {
            // console.log("new item")
            newCart.push({ ...storeItem }) // deep copy as obj
        }
        setCart(newCart)
        localStorage.setItem('cart', JSON.stringify(newCart));
    }
    
    useEffect(() => {
        
        // --jest testing only---
        if(testItems && testItems.length > 0) {
            // console.log(testItems)

            testItems = testItems.map(x => {
                return {
                    ...x,
                    count: 0
                }
            });
            setItems(testItems)
            return
        } 
        // --jest testing only---

        // get database of store items
        fetch(`https://fakestoreapi.com/products/category/women's clothing/`)
            .then(res => res.json())
            .then(res => {
                // console.log(res)

                // add property of count = 0 into each item.
                res = res.map(x => {
                    return {
                        ...x,
                        count: 0
                    }
                });

                setItems(res);
            })

        // get history cart 
        const oldCart = JSON.parse(localStorage.getItem('cart'));
        if (oldCart) {
            setCart(oldCart)
        }

    }, []);

    // console.log(items)

    return (
        <div className={styles.container}>
            <Header cart={cart} currentPage={"store"} handleToggleCart={handleToggleCart} />
            <div className={styles.panel}>
                {items.map(item => <Card key={item.id} item={item} onChange={handleCountChange} addToCart={handleAddToCart} />)}
            </div>
            {showCart && <Cart cart={cart}
                handleToggleCart={handleToggleCart}
                handleCartCountChange={handleCartCountChange}
            />
            }
            <Footer />

        </div>
    )
}

export default Store
