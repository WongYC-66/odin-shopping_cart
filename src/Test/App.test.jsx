import { render, screen } from "@testing-library/react";
import { beforeEach, expect, describe, it, waitFor } from "vitest";
import userEvent from "@testing-library/user-event";
import { BrowserRouter as Router } from 'react-router-dom';
import Home from "../Home/Home";
import Store from "../Store/Store";
import fakeStoreItem from "./fakestore.json"


describe("Home component", () => {

  beforeEach(async () => {
    // Clear mocks and add some testing data after before each test run
    await render(
      <Router>
        <Home />
      </Router>,);
  })

  it("renders Home page correctly", () => {
    expect(screen.getByRole("heading", { name: "THE LONG WEEKEND SALE" })).toBeTruthy()
  });

  it("handles SHOP NOW button click correctly", async () => {
    const user = userEvent.setup();
    const button = screen.getByRole("button", { name: /Shop Now/i });
    await user.click(button);
    expect(screen.getByRole("heading", { name: "THE LONG WEEKEND SALE" }).textContent).toMatch("");
    // expect(screen).toHaveTextContent(/store/);

  });

  it("handles click of navigate button correctly", async () => {
    const user = userEvent.setup();
    const button = screen.getByRole("link", { name: /Store/i });
    await user.click(button);
    expect(screen.getByRole("heading", { name: "THE LONG WEEKEND SALE" }).textContent).toMatch("");
    // expect(screen).toHaveTextContent(/store/);

  });

});

describe("Store component", () => {

  // console.log(fakeStoreItem)
  
  beforeEach(async () => {
    // Clear mocks and add some testing data after before each test run
    await render(
      <Router>
        <Store testItems={fakeStoreItem} />
      </Router>,);
  })

  it("loads the store page with mock items & price correctly", async () => {
    expect(screen.getAllByRole("heading")[2].textContent).toMatch("56.99 $");
  });

  it("add 1 item to cart correctly", async () => {
    const user = userEvent.setup();
    const addOneButton = screen.getAllByRole("button", { name: "+" })[0];
    const addtoCartButton = screen.getAllByRole("button", { name: "Add to Cart" })[0];

    await user.click(addOneButton);
    await user.click(addtoCartButton);
    expect(screen.getByTitle("cartCount").textContent).toMatch("1");
  });


  it("add 1 item to cart correctly (2 pcs)", async () => {
    const user = userEvent.setup();
    const addOneButton = screen.getAllByRole("button", { name: "+" })[0];
    const addtoCartButton = screen.getAllByRole("button", { name: "Add to Cart" })[0];

    await user.click(addOneButton);
    await user.click(addOneButton);
    await user.click(addtoCartButton);
    expect(screen.getByTitle("cartCount").textContent).toMatch("2");
  });

  it("add 1 item to cart correctly (repeatably) ", async () => {
    const user = userEvent.setup();
    const addOneButton = screen.getAllByRole("button", { name: "+" })[0];
    const addtoCartButton = screen.getAllByRole("button", { name: "Add to Cart" })[0];

    await user.click(addOneButton);
    await user.click(addtoCartButton);
    await user.click(addtoCartButton);
    await user.click(addtoCartButton);
    expect(screen.getByTitle("cartCount").textContent).toMatch("3");
  });

  it("add 5 item to cart correctly (each 1 pcs) ", async () => {
    const user = userEvent.setup();
    const addOneButtonGroup = screen.getAllByRole("button", { name: "+" });
    const addtoCartButtonGroup = screen.getAllByRole("button", { name: "Add to Cart" });

    for (let i = 0; i < 5; i++) {
      const addOneButton = addOneButtonGroup[i];
      const addtoCartButton = addtoCartButtonGroup[i];
      await user.click(addOneButton);
      await user.click(addtoCartButton);
    }

    expect(screen.getByTitle("cartCount").textContent).toMatch("5");
  })

  it("loads shopping cart page correctly ", async () => {
    const user = userEvent.setup();
    const showCartButton = screen.getByAltText("StoreLogo");
    await user.click(showCartButton);
    expect(screen.queryByText("Shopping Cart")).not.toBe(null);

  });

  it("closes shopping cart page correctly ", async () => {
    const user = userEvent.setup();
    const showCartButton = screen.getByAltText("StoreLogo");
    await user.click(showCartButton);
    await user.click(showCartButton);
    expect(screen.queryByText("Shopping Cart")).toBe(null);
  });

  it("shows the total amount correctly", async () => {
    const user = userEvent.setup();

    const addOneButtonGroup = screen.getAllByRole("button", { name: "+" });
    const addtoCartButtonGroup = screen.getAllByRole("button", { name: "Add to Cart" });

    for (let i = 0; i < 4; i++) {
      // first 4 item buy 1 each
      const addOneButton = addOneButtonGroup[i];
      const addtoCartButton = addtoCartButtonGroup[i];
      await user.click(addOneButton);
      await user.click(addtoCartButton);
    }
    // fifth item buy 2
    const addOneButton = addOneButtonGroup[4];
    const addtoCartButton = addtoCartButtonGroup[4];
    await user.click(addOneButton);
    await user.click(addOneButton);
    await user.click(addtoCartButton);

    const showCartButton = screen.getByAltText("StoreLogo");
    await user.click(showCartButton);
    expect(screen.getByText("Total : 152.68 $")).toBeInTheDocument();
    //
    
  });

});