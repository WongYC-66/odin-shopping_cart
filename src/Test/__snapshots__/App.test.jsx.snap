// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > Shopping Appls 1`] = `
<div>
  <div
    class="_container_ae5a9c"
  >
    <header>
      <h1>
        YUMEKO
      </h1>
      <div
        class="_navBar_6716bd"
      >
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/store"
          >
            Store
          </a>
        </li>
      </div>
      <div
        class="_icon_6716bd"
      >
        <div
          class="_cartIcon_6716bd"
        >
          <img
            alt="StoreLogo"
            src="/src/Home/store-solid.svg"
          />
        </div>
        <img
          alt="UserLogo"
          src="/src/Home/user-solid.svg"
        />
      </div>
    </header>
    <main>
      <img
        alt="closetPhoto"
        src="/src/Home/closet.webp"
      />
      <h1>
        THE LONG WEEKEND SALE
      </h1>
      <h2>
        Fashion Clothes - Q'3
      </h2>
      <p>
        Grab it while stock last
      </p>
      <button>
        SHOP NOW
      </button>
    </main>
    <footer>
      <p>
        DISCLAIMER - This is a fake shopping website built with React.
      </p>
      Designed and coded by YC Wong @2023
    </footer>
  </div>
</div>
`;

exports[`App component > renders Home page correctly 1`] = `
<div>
  <div
    class="_container_ae5a9c"
  >
    <header>
      <h1>
        YUMEKO
      </h1>
      <div
        class="_navBar_6716bd"
      >
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/store"
          >
            Store
          </a>
        </li>
      </div>
      <div
        class="_icon_6716bd"
      >
        <div
          class="_cartIcon_6716bd"
        >
          <img
            alt="StoreLogo"
            src="/src/Home/store-solid.svg"
          />
        </div>
        <img
          alt="UserLogo"
          src="/src/Home/user-solid.svg"
        />
      </div>
    </header>
    <main>
      <img
        alt="closetPhoto"
        src="/src/Home/closet.webp"
      />
      <h1>
        THE LONG WEEKEND SALE
      </h1>
      <h2>
        Fashion Clothes - Q'3
      </h2>
      <p>
        Grab it while stock last
      </p>
      <button>
        SHOP NOW
      </button>
    </main>
    <footer>
      <p>
        DISCLAIMER - This is a fake shopping website built with React.
      </p>
      Designed and coded by YC Wong @2023
    </footer>
  </div>
</div>
`;
